File: backend/app/core/validation.py
import re
from datetime import date, datetime

from email_validator import EmailNotValidError, validate_email

_UPPERCASE_RE = re.compile(r"[A-Z]")
_DIGIT_RE = re.compile(r"[0-9]")
_SPECIAL_RE = re.compile(r"[^A-Za-z0-9]")
_KEY_RE = re.compile(r"^[a-z][a-z0-9_]*$")


def validate_password_complexity(password: str) -> str:
    """The one password policy for the whole project (docs/modules/authentication.md #4).
    Raises ValueError with a user-facing message on the first rule broken."""
    if len(password) < 8:
        raise ValueError("Password must be at least 8 characters long.")
    if not _UPPERCASE_RE.search(password):
        raise ValueError("Password must contain at least one uppercase letter.")
    if not _DIGIT_RE.search(password):
        raise ValueError("Password must contain at least one number.")
    if not _SPECIAL_RE.search(password):
        raise ValueError("Password must contain at least one special character.")
    return password


def validate_key(value: str) -> str:
    """Shared free-form-string constraint: permissions' module_key/action
    (docs/modules/permissions.md #2) and audit_events' module/action
    (docs/modules/audit_trail.md #2) are both open-ended -- a future
    module registers a new one by writing a row, not editing a Python
    constant -- but still lowercase snake_case so 'Sales' and 'sales'
    can't silently become two different keys."""
    if not _KEY_RE.match(value):
        raise ValueError("must be lowercase snake_case, e.g. 'sales' or 'view'")
    return value


def check_max_length(value: str, max_length: int, field_label: str) -> str:
    """Enforces a field's DB column width at the API boundary, checked
    after whatever normalization (strip, case-fold, digit-extraction)
    the caller already applied -- so an over-limit value is rejected
    with a clear 422 instead of being silently truncated (SQLite, or
    MySQL outside strict mode) or raising a raw IntegrityError/DataError
    at the database layer (docs/modules/common_validation.md)."""
    if len(value) > max_length:
        raise ValueError(f"{field_label} must be at most {max_length} characters.")
    return value


def normalize_email(email: str) -> str:
    """One normalization for every place an email is stored or compared
    -- lowercase, trimmed -- so 'Ada@Example.com' and 'ada@example.com'
    can't silently become two different users or bypass a duplicate
    check (docs/modules/common_validation.md #1)."""
    return email.strip().lower()


def validate_email_format(email: str) -> str:
    """Format checking via the `email_validator` package already used by
    Pydantic's `EmailStr` (see app/schemas/user.py) -- reused here rather
    than duplicated as a hand-rolled regex, for the call sites (a future
    signup/user-creation flow) that need a plain callable rather than a
    Pydantic field type. Raises ValueError with a user-facing message;
    never leaks the library's own exception type/message verbatim."""
    try:
        validate_email(email, check_deliverability=False)
    except EmailNotValidError as exc:
        raise ValueError(f"'{email}' is not a valid email address.") from exc
    return normalize_email(email)


def validate_company_email_domain(email: str, allowed_domain: str | None) -> str:
    """Only the organisation's configured company domain is allowed --
    but the domain is configuration (Organisation.email_domain), not a
    hard-coded constant, so this takes it as a parameter rather than
    reading global settings. An organisation that hasn't configured one
    (allowed_domain is None) has no domain restriction at all."""
    if allowed_domain is None:
        return normalize_email(email)
    normalized = normalize_email(email)
    domain = normalized.rsplit("@", 1)[-1]
    if domain != allowed_domain.strip().lower():
        raise ValueError(f"Email must use the '{allowed_domain}' company domain.")
    return normalized


def validate_date_range(
    start: date | datetime,
    end: date | datetime,
    *,
    start_label: str = "Start date",
    end_label: str = "End date",
) -> None:
    """The one `from <= to` comparison mechanism (docs/modules/common_validation.md
    #2) -- equal dates are valid (same-day validity), only start > end is
    rejected. Whether past/future dates are allowed at all is a business
    rule the calling module enforces itself; this only orders the pair."""
    if start > end:
        raise ValueError(f"{start_label} must not be after {end_label}.")

File: frontend/src/lib/validation.ts
/** Mirrors backend/app/core/validation.py and id_formats.py -- the
 * frontend's copy of the same mechanisms, for UI-side validation (zod
 * schemas, inline field checks) ahead of the server's authoritative
 * check (docs/modules/common_validation.md core principle: "Frontend
 * validation improves usability; server-side validation enforces
 * correctness."). Every function here throws/returns the same shape a
 * caller can drop straight into a zod `.refine()` or `superRefine()`. */

export function normalizeEmail(email: string): string {
  return email.trim().toLowerCase()
}

/** Only the organisation's configured company domain is allowed -- but
 * the domain is configuration (Organisation.email_domain), not
 * hard-coded, so it's a parameter. `allowedDomain` of null/undefined
 * means the organisation has no domain restriction. */
export function validateCompanyEmailDomain(email: string, allowedDomain: string | null | undefined): string {
  const normalized = normalizeEmail(email)
  if (!allowedDomain) return normalized
  const domain = normalized.split('@').pop() ?? ''
  if (domain !== allowedDomain.trim().toLowerCase()) {
    throw new Error(`Email must use the '${allowedDomain}' company domain.`)
  }
  return normalized
}

/** The one `from <= to` comparison mechanism (docs/modules/common_validation.md
 * #2) -- equal values are valid (same-day validity), only start > end is
 * rejected. Accepts ISO date/datetime strings, as produced by
 * `<input type="date">`/`<input type="datetime-local">`
 * (components/forms/DateField.tsx), or Date objects. Whether past/future
 * dates are allowed at all is left to the calling module. */
export function validateDateRange(
  start: string | Date,
  end: string | Date,
  { startLabel = 'Start date', endLabel = 'End date' }: { startLabel?: string; endLabel?: string } = {},
): void {
  const startTime = start instanceof Date ? start.getTime() : new Date(start).getTime()
  const endTime = end instanceof Date ? end.getTime() : new Date(end).getTime()
  if (startTime > endTime) {
    throw new Error(`${startLabel} must not be after ${endLabel}.`)
  }
}

export interface IdFormat {
  pattern: RegExp
  example: string
}

/** One prefix+digit-count shape per entity, shared with backend's
 * app/core/id_formats.py -- Quotation (QXXXXXX), Order (OXXXXXX), User
 * (XXXXX), Product (PRXXXX), Material (MXXXX). */
export const ID_FORMATS = {
  quotation: { pattern: /^Q\d{6}$/, example: 'QXXXXXX' },
  order: { pattern: /^O\d{6}$/, example: 'OXXXXXX' },
  user: { pattern: /^\d{5}$/, example: 'XXXXX' },
  product: { pattern: /^PR\d{4}$/, example: 'PRXXXX' },
  material: { pattern: /^M\d{4}$/, example: 'MXXXX' },
} as const satisfies Record<string, IdFormat>

export type IdFormatName = keyof typeof ID_FORMATS

export function validateIdFormat(name: IdFormatName, value: string): string {
  const format = ID_FORMATS[name]
  if (!format.pattern.test(value)) {
    throw new Error(`must match the format ${format.example}`)
  }
  return value
}
